### PR TITLE
Add learn-weights CLI step with PU learning and Optuna tuning

### PR DESCRIPTION
**Pull request by iankchristie**
Introduces a new `reVeal learn-weights` pipeline step that trains a PUExtraTrees model on a normalized grid using point labels as positive samples. Derives feature importance weights for use with `score-weighted`.

- Add reVeal/pu/ — PUExtraTree/PUExtraTrees adapted from https://github.com/jonathanwilton/PUExtraTrees, with modifications from /projects/largeload/.../models/PUExtraTrees/ (deterministic seeding, joblib parallelism)
- Add reVeal/learn_weights.py — data preparation logic adapted from /projects/largeload/.../models/prepare_data.py (DataHandler), training and metrics from /projects/largeload/.../models/models.py (ModelTrainer)
- Add Optuna-based hyperparameter tuning for class_prior — adapted from ModelTrainer.tune_hyperparameters() and ModelTrainer.objective() in /projects/largeload/.../models/models.py
- Add CLI command (reVeal/cli/learn_weights.py), config (reVeal/config/learn_weights.py), and tests — new code following existing reVeal CLI patterns (score_weighted, normalize)
- Add joblib, scipy, optuna to pyproject.toml and environment.yml

Tested by running the new CLI tool like so
```
python -m reVeal.cli.cli learn-weights -c /projects/largeload/geospatial/runs/test_scenarios/learned_weights_2026-05-19_agg64/config_learn_weights.json
```

You can use the learned_weights_2026-05-19_agg64/ directory to test this code.

Next Steps:
This works well to generate the score weights. However, in practice there many of these features are highly correlated or we want to manually exclude for other purposes. We want a few feature engineering tools that will allow users to iterate on the score outputs. This will include:
1. An exclude list
2. Visualization features (dendrogram and correlation matrix) to help users feature engineer before it goes into the score_weighted step.

### CODE REVIEW

File: reVeal/learn_weights.py [hunk 1]
```diff
@@ -0,0 +1,495 @@
+"""
+learn_weights module - Core logic for PU learning-based weight derivation.
+
+Trains a PUExtraTrees model on a normalized grid using point labels as positive
+samples and auto-sampled background cells as unlabeled samples. Produces a
+score-weighted configuration with feature importances normalized as weights.
+"""
+import json
+import logging
+from pathlib import Path
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/learn_weights.py]: DOING

File: reVeal/learn_weights.py [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+
+
+def prepare_pu_data(
+    grid_df,
+    labels_gdf,
+    attributes,
+    background_samples=10000,
+    test_size=0.2,
+    validation_size=0.1,
+    random_state=42,
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/learn_weights.py:28]: DOING

File: reVeal/learn_weights.py [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+    # Train model
+    model = PUExtraTrees(
+        n_estimators=n_estimators,
+        risk_estimator="nnPU",
+        n_jobs=n_jobs,
+        random_state=random_state,
+    )
+    logger.info(f"Training PUExtraTrees with {n_estimators} trees...")
+    model.fit(P=p_tr, U=u_tr, pi=class_prior)
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/learn_weights.py:181]: DOING

File: reVeal/learn_weights.py [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+        u_test = g[g["gid"].isin(data_splits["background_test_gids"])][
+            attributes
+        ].to_numpy()
+        x_test = np.vstack((p_test, u_test))
+        y_test = np.concatenate((
+            np.ones(len(p_test), dtype=int),
+            np.zeros(len(u_test), dtype=int),
+        ))
+        y_pred = (model.predict(x_test) == 1).astype(int)
+        metrics["auc"] = float(roc_auc_score(y_test, y_pred))
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/learn_weights.py]: DOING

File: reVeal/learn_weights.py [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+
+    x_val = np.vstack((p_val, u_val))
+    y_val = np.concatenate((
+        np.ones(len(p_val), dtype=int),
+        np.zeros(len(u_val), dtype=int),
+    ))
+    y_pred = (model.predict(x_val) == 1).astype(int)
+
+    if metric == "auc":
+        return float(roc_auc_score(y_val, y_pred))
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/learn_weights.py]: DOING

File: reVeal/config/learn_weights.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+"""
+config.learn_weights module - Configuration for learn-weights command.
+"""
+from typing import List, Literal, Optional
+
+from pydantic import model_validator, FilePath, Field
+from typing_extensions import Annotated
+
+from reVeal.config.config import BaseModelStrict, BaseGridConfig
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/config/learn_weights.py]: DOING

File: tests/test_learn_weights.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
+# -*- coding: utf-8 -*-
+"""
+Tests for learn_weights module.
+"""
+import json
+import numpy as np
+import geopandas as gpd
+import pytest
+from shapely.geometry import Point, box
+
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply tests/test_learn_weights.py]: DOING

File: pyproject.toml [hunk 8]
```diff
@@ -33,14 +33,17 @@ dependencies = [
...
+  "joblib>=1.4.0,<2",
   "libpysal>=4.13.0,<5",
   "NLR-GAPs>=0.9.1,<1",
   "NLR-rex>=0.5.0,<1",
+  "optuna>=4.0.0,<5",
   "pandas>=2.3.3,<3",
   "pyarrow>=22.0.0,<23",
   "pydantic>=2.12.5,<3",
   "pyogrio>=0.12.1,<0.13",
   "rasterio>=1.4.4,<2",
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply pyproject.toml:45]: DOING

File: environment.yml [hunk 9]
```diff
@@ -8,11 +8,14 @@ dependencies:
...
     - pyogrio>=0.12.1,<0.13
     - pyarrow>=22.0.0,<23
     - proj-data>=1.24,<2
+    - scipy>=1.14.0,<2
     - tqdm>=4.67.1,<5
     - exactextract>=0.2.2,<1
+    - optuna>=4.0.0,<5
     - pydantic>=2.12.5,<3
     - libpysal>=4.13.0,<5
     - pip>=25.2
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply environment.yml:22]: DOING

File: reVeal/pu/trees.py [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+class PUExtraTrees:
+    def __init__(self, n_estimators = 100,
+                 risk_estimator = 'nnPU',
+                 loss = 'quadratic',
+                 max_depth = None,
+                 min_samples_leaf = 1,
+                 max_features = 'sqrt',
+                 max_candidates = 1,
+                 n_jobs = 1,
+                 random_state = 42):
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/pu/trees.py:17]: Ignoring

File: reVeal/pu/tree.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+import scipy.sparse
+
+class PUExtraTree:
+    def __init__(self, risk_estimator = "nnPU",
+                 loss = "quadratic",
+                 max_depth = None, 
+                 min_samples_leaf = 1, 
+                 max_features = "sqrt", 
+                 max_candidates = 1,
+                 random_state = None):
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/pu/tree.py:14]: Ignoring

File: reVeal/pu/trees.py [hunk 12]
```diff
@@ -0,0 +1,463 @@
+# PU ExtraTrees - A Random Forest Classifier for PU Learning
+# Adapted from https://github.com/jonathanwilton/PUExtraTrees
+from .tree import PUExtraTree
+from joblib import Parallel, delayed
+import scipy
+import numpy as np
+
+class PUExtraTrees:
```
[the remark this answers is not in the document]
Comment (iankchristie) [reply reVeal/pu/trees.py:8]: Ignoring

**Review by ppinchuk — APPROVED** [review]
Thanks for adding tests and keeping the environment lockfile updated!
